the-collab-lab/tcl-46-smart-shopping-list
# Estimate purchase date upon item purchase (user checks off)

## Description

Added logic (in .utils, updateItem() in `firebase.js`, and inside the useEffect in `ListItem.js`) so that when the user checks off an item as purchased, the app can calculate and store a guess for when the item should be purchased next. 

The user does not currently see any changes to the UI, but the dateNextPurchased is updated in Firebase to enable future functionality.

### Considerations
- Avoiding infinite loop or double-counting: The app behavior seems to vary despite running checks on the same code/sub-branches. Excluding `isChecked` from the first useEffect deps inside `ListItem.js` seems to remove all unexpected behavior, but inclusion of `isChecked` in deps seems to work fine on other browser/device? **Needs further confirmation in this review**. 
- Based on `calculateEstimate` docs, prevEst (previous estimate) is left undefined when no dateLastPurchased is on record. 
- As well, daysSinceLastPurchase is calculated as time elapsed since item added (dateCreated) and current time. A buffer of minimum 1 day was added to avoid a recommendation of immediate/same-day prediction based on the current calculations.
- Inconsistent reference at times in the docs to the "days elapsed" param - would be open to standardizing to daysSinceLastTransaction (since that is more inclusive of the interval values passed for that param).

## Related Issue

Closes #10 

## Acceptance Criteria

## Type of Changes



|     | Type                       |
| --- | -------------------------- |
|     | :bug: Bug fix              |
| ✓   | :sparkles: New feature     |
|     | :hammer: Refactoring       |
|     | :100: Add tests            |
|     | :link: Update dependencies |
|     | :scroll: Docs              |

## Updates
No UI changes.

## Testing Steps / QA Criteria
1. Pull and check out `yh-hy-estimatePurchaseDate-10` and run with `npm start`. In a separate window, open the [Firebase console](https://console.firebase.google.com/u/1/project/tcl-46-smart-shopping-list/firestore/data/~2Fbean%20selma%20wade~2FAzUvhs9KvR4L9PePSORF). Join `my test list` and in Firebase, navigate to an item you recognize from the app List page for the next steps.
2.  Check off the item in the app and see the updates to the Firebase doc in real time for dateLastPurchased, dateNextPurchased, and totalPurchases.
  - If totalPurchases is less than 2, the dateNextPurchased should correspond to the difference between dateLastPurchased and dateCreated. If totalPurchases is greater than 2, dateNextPurchased should provide a date that reflects the difference between dateLastPurchased and the previous (pre-update) dateNextPurchased estimate.
  - totalPurchases should increment by 1 only each time isChecked is true, and isChecked status should be synced to the live app view for that item.
